File: pspiHost/pspiHost.cpp
#include "stdafx.h"
#include "pspiHost.h"
#include "TPspiCore.h"
TPspiCore piCore;
static int imgSclIndex, maskSclIndex;
static char version[4];
//-----------------------------------------------------------------
// get version
//-----------------------------------------------------------------
const char* pspiGetVersion(void)
{  
	memcpy(version, "0.9\0", 4);
	return version;
}
//-----------------------------------------------------------------
// set path
//-----------------------------------------------------------------
int pspiSetPath(const wchar_t *filterFolder)
{
	return piCore.SetPath(filterFolder);
}
//---------------------------------------------------------------------------
// set filter roi
//---------------------------------------------------------------------------
int pspiSetRoi(int top, int left, int bottom, int right)
{
	piCore.HostRecord.roiRect.top = top;
	piCore.HostRecord.roiRect.left = left;
	piCore.HostRecord.roiRect.bottom = bottom;
	piCore.HostRecord.roiRect.right = right;
	piCore.HostRecord.roiRect.Normalize();
	return 0;
}
//-----------------------------------------------------------------
// set imege orientatio - As is or invert scanlines (this applies to mask also)
//-----------------------------------------------------------------
int pspiSetImageOrientation(TImgOrientation orientation)
{
	piCore.HostRecord.imgOrientation = orientation;
	return 0;
}
//-----------------------------------------------------------------
// set working image - pass contiguous buffer
//-----------------------------------------------------------------
int pspiSetImage(TImgType type, int width, int height, void *imageBuff, int imageStride, void *alphaBuff, int alphaStride)
{
	pspiSetRoi();	// clear roi
	return piCore.SetImage(type, width, height, imageBuff, imageStride, alphaBuff, alphaStride);
}
//-----------------------------------------------------------------
// set working mask - pass contiguous buffer (grayscale image)
//-----------------------------------------------------------------
int pspiSetMask(int width, int height, void *maskBuff, int maskStride, bool useMaskByPi)
{
	return piCore.SetMask(width, height, maskBuff, maskStride, useMaskByPi);
}
//-----------------------------------------------------------------
// start working image creation - for passing scanlines
//-----------------------------------------------------------------
int pspiStartImageSL(TImgType type, int width, int height, bool externalAplha)
{
	pspiSetRoi();	// clear roi
	imgSclIndex = 0;
	return piCore.StartImageSL(type, width, height, externalAplha);
}
//-----------------------------------------------------------------
// add image scanline (use it in loop)
//-----------------------------------------------------------------
int pspiAddImageSL(void *imageScanLine, void *alphaScanLine)
{
	return piCore.AddImageSL(imgSclIndex++, imageScanLine, alphaScanLine);
}
//-----------------------------------------------------------------
// finish image creation using scanlines
//-----------------------------------------------------------------
int pspiFinishImageSL(int imageStride, int alphaStride)
{
	return piCore.FinishImageSL(imageStride, alphaStride);
}
//-----------------------------------------------------------------
// start mask creation - for passing scanlines
//-----------------------------------------------------------------
int pspiStartMaskSL(int width, int height, bool useMaskByPi)
{
	maskSclIndex = 0;
	return piCore.StartMaskSL(width, height, useMaskByPi);
}
//-----------------------------------------------------------------
// add mask scanline (use it in loop)
//-----------------------------------------------------------------
int pspiAddMaskSL(void *maskScanLine)
{
	return piCore.AddMaskSL(maskSclIndex++, maskScanLine);
}
//-----------------------------------------------------------------
// finish mask creation using scanlies
//-----------------------------------------------------------------
int pspiFinishMaskSL(int maskStride)
{
	return piCore.FinishMaskSL(maskStride);
}
//-----------------------------------------------------------------
// set progress procedure call back
//-----------------------------------------------------------------
int pspiReleaseAllImages(void)
{
	pspiSetRoi();	// clear roi
	piCore.ReleaseAllImages();
	return 0;
}
//-----------------------------------------------------------------
// set progress procedure call back
//-----------------------------------------------------------------
int pspiSetProgressCallBack(PROGRESSCALLBACK progressProc)
{
	piCore.ProgressCallBack = progressProc;
	return 0;
}
//-----------------------------------------------------------------
// set color picker procedure call back
//-----------------------------------------------------------------
int pspiSetColorPickerCallBack(COLORPICKERCALLBACK colorPickerProc)
{
	piCore.ColorPickerCallback = colorPickerProc;
	return 0;
}
//-----------------------------------------------------------------
// Load plugin
//-----------------------------------------------------------------
int pspiPlugInLoad(const wchar_t *filter)
{
	return piCore.PlugInLoad(filter);
}
//-----------------------------------------------------------------
// Show plug-in about
//-----------------------------------------------------------------
int pspiPlugInAbout(HWND hWnd)
{
	return piCore.PlugInAbout(hWnd);
}
//-----------------------------------------------------------------
// Execute loaded plug-in
//-----------------------------------------------------------------
int pspiPlugInExecute(HWND hWnd)
{
	return piCore.PlugInExecute(hWnd);
}
//---------------------------------------------------------------------------
// Enumerate Filters from directory using call back function
//---------------------------------------------------------------------------
int pspiPlugInEnumerate(ENUMCALLBACK enumFunc, bool recurseSubFolders)
{
	if (enumFunc == NULL)
		return PSPI_ERR_FILTER_BAD_PROC;
	return piCore.PlugInEnumerate(enumFunc, recurseSubFolders);
}

File: Additional/Delphi/pspiHost.pas
unit pspiHost;

interface

uses pspiGlobals, Windows;


type
  int = Integer;

// get psipHost version
function pspiGetVersion(): PAnsiChar cdecl;

// set path
function pspiSetPath(filterFolder: Pointer): int cdecl;

// set roi
function pspiSetRoi(top: int = 0; left: int = 0; bottom: int = 0; right: int = 0): int  cdecl;

// set image orientation
function pspiSetImageOrientation(orientation: TImgOrientation): int cdecl;

// set image using contiguous memory buffer pointer
// note: source image is shared - do not destroy source image in your host program before plug-in is executed
function pspiSetImage(imgtype: TImgType; width: int; height: int; imageBuff: Pointer; imageStride: int; alphaBuff: Pointer = nil; alphaStride: int = 0): int cdecl;

// set mask using contiguous memory buffer pointer
// note: source mask is shared - do not destroy source maske in your host program before plug-in is executed
function pspiSetMask(width: int = 0; height : int = 0; maskBuff: Pointer = nil; maskStride: int = 0;  useMaskByPi: Boolean = true): int cdecl;

// block for adding image scanlines (possibly non-contiguous image)
// note: source image is shared - do not destroy source image in your host program before plug-in is executed
function pspiStartImageSL(imgtype: TImgType; width: int; height: int; externalAlpha: Boolean = false): int  cdecl;
function pspiAddImageSL(imageScanLine: Pointer; alphaScanLine: Pointer = nil): int cdecl;
function pspiFinishImageSL(imageStride: int = 0; alphaStride: int = 0): int  cdecl;

// block dor addding mask scanlines (possibly non-contiguous mask)
// note: source mask is shared - do not destroy source maske in your host program before plug-in is executed
function pspiStartMaskSL(width: int; height: int; useMaskByPi: Boolean = true): int cdecl;
function pspiAddMaskSL(maskScanLine: Pointer): int cdecl;
function pspiFinishMaskSL(maskStride: int = 0): int cdecl;

// release all images - all image buffers including mask are released when application is closed; but sometimes it's necessary to free memory (big images)
function pspiReleaseAllImages(): int cdecl;

// set call backs of interset
function pspiSetProgressCallBack(progressProc: TPROGRESSCALLBACK): int cdecl;
function pspiSetColorPickerCallBack(colorPickerProc: TCOLORPICKERCALLBACK): int cdecl;

// plug-in related
function pspiPlugInLoad(filter: Pointer): int cdecl;
function pspiPlugInAbout(hWnd: HWND = 0): int cdecl;
function pspiPlugInExecute(hWnd: HWND = 0): int cdecl;
function pspiPlugInEnumerate(enumFunc: TENUMCALLBACK; recurseSubFolders: Boolean = true): int cdecl;

implementation

function pspiGetVersion(): PAnsiChar; external 'pspiHost.dll';
function pspiSetPath(filterFolder: Pointer): int; external 'pspiHost.dll';
function pspiSetRoi(top: int = 0; left: int = 0; bottom: int = 0; right: int = 0): int; external 'pspiHost.dll';
function pspiSetImageOrientation(orientation: TImgOrientation): int; external 'pspiHost.dll';
function pspiSetImage(imgtype: TImgType; width: int; height: int; imageBuff: Pointer; imageStride: int; alphaBuff: Pointer = nil; alphaStride: int = 0): int; external 'pspiHost.dll';
function pspiSetMask(width: int = 0; height: int = 0; maskBuff: Pointer = nil; maskStride: int = 0;  useMaskByPi: Boolean = true): int; external 'pspiHost.dll';
function pspiStartImageSL(imgtype: TImgType; width: int; height: int; externalAlpha: Boolean = false): int; external 'pspiHost.dll';
function pspiAddImageSL(imageScanLine: Pointer; alphaScanLine: Pointer = nil): int; external 'pspiHost.dll';
function pspiFinishImageSL(imageStride: int = 0; alphaStride: int = 0): int; external 'pspiHost.dll';
function pspiStartMaskSL(width: int; height: int; useMaskByPi: Boolean = true): int; external 'pspiHost.dll';
function pspiAddMaskSL(maskScanLine: Pointer): int; external 'pspiHost.dll';
function pspiFinishMaskSL(maskStride: int = 0): int; external 'pspiHost.dll';
function pspiReleaseAllImages(): int; external 'pspiHost.dll';
function pspiSetProgressCallBack(progressProc: TPROGRESSCALLBACK): int; external 'pspiHost.dll';
function pspiSetColorPickerCallBack(colorPickerProc: TCOLORPICKERCALLBACK): int; external 'pspiHost.dll';
function pspiPlugInLoad(filter: Pointer): int; external 'pspiHost.dll';
function pspiPlugInAbout(hWnd: HWND = 0): int; external 'pspiHost.dll';
function pspiPlugInExecute(hWnd: HWND = 0): int; external 'pspiHost.dll';
function pspiPlugInEnumerate(enumFunc: TENUMCALLBACK; recurseSubFolders: Boolean = true): int; external 'pspiHost.dll';

end.

